Java annotation array parsing fixes

retronym opened the pull request:
- support trailing commas
  - support empty array initializer lists
  - when we hit syntax we don't support within an array initializer,
    brace match our way to the end of `}` that concludes the
    list.

retronym committed:
Java annotation array parsing fixes

  - support trailing commas
  - support empty array initializer lists
  - when we hit syntax we don't support within an array initializer,
    brace match our way to the end of `}` that concludes the
    list.

dwijnand commented on test/files/neg/java-annotation-match-error/J.java:
```diff
@@ -0,0 +1,11 @@
+  })
+  public static void foo() {}
+
+  int int int
```
Is this meant to be part of the test?

dwijnand commented on src/compiler/scala/tools/nsc/javac/JavaParsers.scala:
```diff
@@ -365,12 +365,15 @@ trait JavaParsers extends ast.parser.ParsersCommon with JavaScanners {
             case _ if in.token == IDENTIFIER =>
               qualId(orClassLiteral = true)
+            case _ =>
+              syntaxError(in.currentPos, "Annotation value expected", skipIt = true)
```
I think following our reasoning on how non-constant values weren't supported before and aren't supported still, we should continue to just skip those, no?  No error?

lrytz replied:
Yeah, the syntax allows a lot more https://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html#jls-ElementValue

For example, this should not give a syntax error

```
public class A {
  public @interface Ann { int value(); } 
  @Ann((1 == 2) ? 3 : 4)
  public static void foo() { return; }
}
```

retronym replied:
I'll amend and add more neg tests. The existing bailout mechanism doesn't deal with the unsupported syntax within array initializers.

retronym replied:
I also found an existing bug with annotation top level classes in the empty package:

```java
@Ann(values = {"foo", "bar"})
class Test {}

@interface Ann {
    String[] values();
}
```

```
sandbox/Test.java:2: error: `package' expected but `class' found.
class Test {}
^
sandbox/Test.java:6: error: identifier expected but eof found.
}
 ^
two errors found
```

Another edge case arises when trailing commas are uses in array initializers:

```java
package p1;

@Ann(values = {1, 2, })
class Test {}

@interface Ann {
    int[] values();
}
```

That's legal but gives a `MatchError` in 2.12.x (like the main test case in this PR). My current patch drops the arguments unnecessarily. I'll refine it to support the trailing comma. We're in good company though, `javac` also rejected the trailing commas [until 2012](https://bugs.java.com/bugdatabase/view_bug.do?bug_id=6337964)!

retronym committed:
Avoid slow patmat exhaustiveness analysis in a test

Was:
```
ok 79 - run/t8928                               [duration 188.25s]
```

Now:

```
ok 1 - run/t8928                               [duration 9.60s]
```

retronym commented:
/rebuild

dwijnand reviewed (approved):
Oh, wow.  Thanks for the deep surgery, the docs and the other test variants!
